File: elixir/word-count/word_count.exs
defmodule Words do
  def count(sentence) do
    sanitize(sentence)
    |> String.split
    |> Enum.reduce(%{}, fn(word, acc) ->
      count = Map.get(acc, word, 0)

      Map.put(acc, word, count + 1)
    end)
  end

  def sanitize(sentence) do
    String.trim(sentence)
    |> String.downcase
    |> String.replace(~r/[^\p{L}0-9-]/u, " ")
  end
end

File: erlang/word-count/src/word_count.erl
-module(word_count).

-export([count/1, test_version/0]).

count(String) ->
  count(words(String), dict:new()).

count([], Result) ->
  Result;

count([Word | Rest], Result) ->
  count(Rest, dict:update_counter(Word, 1, Result)).

words(String) ->
  string:tokens(normalize(String), " ").

normalize(String) ->
  Lowered = string:to_lower(String),
  re:replace(Lowered, "[^a-z0-9]", " ", [global, {return, list}]).


test_version() -> 1.
